src/main/java/com/google/ehub/utility/UserRecommendationUtils.java

```diff
@@ -15,18 +17,60 @@
 
   /**
    * Finds the most recommended emails up to a maximum of ten recommendations in descending order.
+   * In the case of a tie, the order will use increasing lexicographical order.
    *
    * @param itemLikes map with the key representing an itemId and the value representing the emails
    *     that liked that item
    * @return list containing the most recommended emails in descending order
    */
   public List<String> getRecommendedEmails(Map<Long, Set<String>> itemLikes) {
+    return getRecommendedEmailsInDescendingOrder(getEmailMinHeap(getEmailFreqs(itemLikes)));
+  }
+
+  private Map<String, Integer> getEmailFreqs(Map<Long, Set<String>> itemLikes) {
+    Map<String, Integer> emailFreqs = new HashMap<String, Integer>();
+
+    for (Map.Entry<Long, Set<String>> emailGroup : itemLikes.entrySet()) {
+      for (String email : emailGroup.getValue()) {
+        emailFreqs.merge(email, 1, (oldFreq, deltaFreq) -> oldFreq + deltaFreq);
+      }
+    }
+
+    return emailFreqs;
+  }
+
+  private PriorityQueue<Map.Entry<String, Integer>> getEmailMinHeap(
+      Map<String, Integer> emailFreqs) {
+    PriorityQueue<Map.Entry<String, Integer>> emailMinHeap = new PriorityQueue<>((a, b) -> {
+      // If the counts are tied, then compare emails lexicographically.
+      if (a.getValue() == b.getValue()) {
+        return b.getKey().compareTo(a.getKey());
+      }
+
+      return Integer.compare(a.getValue(), b.getValue());
+    });
+
+    for (Map.Entry<String, Integer> entry : emailFreqs.entrySet()) {
+      emailMinHeap.offer(entry);
+
+      if (emailMinHeap.size() > MAX_NUMBER_OF_RECOMMENDATIONS) {
+        emailMinHeap.poll();
+      }
+    }
+
+    return emailMinHeap;
+  }
+
+  private List<String> getRecommendedEmailsInDescendingOrder(
+      PriorityQueue<Map.Entry<String, Integer>> emailMinHeap) {
     List<String> recommendedEmails = new ArrayList<String>();
 
-    // TODO: Implement algorithm that finds the users that have the highest amount of common items,
-    // could be implemented
-    // by using a Map<String, int> storing counts and a priority queue to get the
-    // "maxNumberOfRecommendations"
+    while (!emailMinHeap.isEmpty()) {
+      recommendedEmails.add(emailMinHeap.poll().getKey());
+    }
+
+    // The most recommended emails should come first.
+    Collections.reverse(recommendedEmails);
 
     return recommendedEmails;
   }
```

Comment:
Nice job with the algorithm 😄 

I think Guava's [MinMaxPriorityQueue](https://guava.dev/releases/22.0/api/docs/com/google/common/collect/MinMaxPriorityQueue.html) can be very useful here. It automatically removes the last ("greatest") element if the PQ reaches the MAX size. With this behavior, you can just reverse the comparator that you created for the PQ. Then you won't have to reverse the items at the end to get them in descending order, since they will already be in the correct order!

Reply:
Thanks for the suggestion that shortened the code a little more.